### PR TITLE
Feature: Add set up development environments docs

### PR DESCRIPTION
**Pull request by syphax-bouazzouni**
### Context

When we have new developers in the Alliance that don't just want to run the appliance but also want to do some development and enhancement to their local installation. 

The problem is that it's hard to start a local development environment and understand how our code works.

This PR tries to give a draft version of two documentation pages:

1. Ontoportal Architecture: That describes our architecture, you can see a preview [here](https://ontoportal-lirmm.github.io/documentation/docs/developers-docs/global_architecture/) 
2. Set up a development environment: That gives two guides on how to set up a development environment for the front and backend; you can see a preview [here](https://ontoportal-lirmm.github.io/documentation/docs/developers-docs/development-environment/)

### Changes

- Update getting started content(https://github.com/ontoportal/documentation/pull/22/commits/b556b4bde1103295b7d7c3841cdb581831efea15)
- Add backend environment documentation page (https://github.com/ontoportal/documentation/pull/22/commits/8e0003875e6f8f29e515fb588c6524e6b96dba03)
- Add frontend environment documentation page(https://github.com/ontoportal/documentation/pull/22/commits/185b9bc91514497dcb996da772d1a1b9c5add5f2)
- Update and move the architecture page(https://github.com/ontoportal/documentation/pull/22/commits/a762b3bae1d0efd9f37acfba364216bbe8e87f47)
- Rename the old getting_started to debugging_ontoportal (https://github.com/ontoportal/documentation/pull/22/commits/d7385907f082912e85d5b6dd076d8b2b9bcb29eb)
- Update the developer section title (https://github.com/ontoportal/documentation/pull/22/commits/a606ce16a6166344e5d3488e033c9afc536ccfd4)

### CODE REVIEW

**Comment by syphax-bouazzouni** [comment]
@RaimiInfai here are the docs that we spoke about. As you are the ones currently setting up a new portal (BioDivPortal), your feedback is welcomed.

Same for you @Bouchemel-Nasreddine; hope this will help you set up your dev environment for your internship in the Industry portal.

And for the ones that want to edit or add something to this PR, collaboration is open just add your commits in the branch [feature/add-development-environment-guides](https://github.com/ontoportal/documentation/tree/feature/add-development-environment-guides)

**Review by RaimiInfai — APPROVED** [review]

**Review by stdotjohn — APPROVED** [review]
I think that this pull request should be approved. In fact I think that much of the approval process could be simplified for this type of pull request, anyone who goes to the trouble of doing a writeup should be allowed to add it.  I think that the use of VirtualBox
 is very interesting.  Qemu utilities, for example, should allow us to convert such a VM to other platforms such as KVM or VMWare. I am interested in Docker at the moment not because it can be lightweight (my containers probably are not) but because such VMs are defined by text files that define their history.
 
My major confusion about this document is that at tne end it dafines how to run the target executable not how to "run" it "in" a ruby IDE  (with, for example, its nice graphical representation of breakpoints) such as RubyMine.  In particular, debugging a container containing a ruby executable has challenges (remote debug and/or sharing a volume with the host) which I am investigating now.

**Comment by syphax-bouazzouni** [comment]
> I think that this pull request should be approved. In fact, I think that much of the approval process could be simplified for this type of pull request, anyone who goes to the trouble of doing a writeup should be allowed to add it. 

The pull request approval process may slow down the contributions. But it's important to be transparent about our changes and let (all related) people know about it.

But yeah, maybe we can improve it, firstly to not wait for the approval of everyone, and just have a limited number of approval before merging (e.g 2 approval per PR). (we will see in the future, but for now, it doesn't bother me to wait) 

> My major confusion about this document is that at the end it defines how to run the target executable not how to "run" it "in" a ruby IDE (with, for example, its nice graphical representation of breakpoints) such as RubyMine. In particular, debugging a container containing a ruby executable has challenges (remote debugging and/or sharing a volume with the host) which I am investigating now.

The problem with IDEs like RubyMine is that it is not free and that everyone has their preferred IDE (most people use for example VS Code because it's free and lightweight).

So here I tried to make something, that will work everywhere for free.

**Review by graybeal — APPROVED** [review]
Does the existence of passwords (and perhaps other config details) in this documentation (in the config file) create a security issue that can be mitigated in any way?

**Comment by graybeal** [comment]
I agree with @stdotjohn in that pull requests for documentation can be assumed to be 'likely better than what came before', if we believe in best intentions of the submitter. For example I found 2 typos in this one but assessed that it was far better to install it as is, than to not get my approval or delay merging because I didn't have time to recommend those changes. And small issues are easy to fix later. So if the first reviewer is doing a reasonably thorough review, and approves, I think that is enough; the rest of the reviews are gravy. 

But if you merge a pull request after m reviews, do the n-m reviewers who are left still have the option to review it? Ideally they would have the option and would exercise it, because yes the dev teams should have an idea what is changing 'under them'.

**Comment by stdotjohn** [comment]
> Does the existence of passwords (and perhaps other config details) in this documentation (in the config file) create a security issue that can be mitigated in any way?

I don't really think that this poses a development challenge.  First of all, for an attacker to get onto a development machine   at all is a serious security breach.  A typical. way that this could happen is when an attacker sends the victim an attachment that the victim clicks on.   (There are many other ways.). A developer depends on NAT and firewalls and other such mechanisms to protect herself.

In deployment such issues are very serious however.

File: docs/developers-docs/development-environment/backend_environment.md
Comment (RaimiSol) [inline docs/developers-docs/development-environment/backend_environment.md:47]: Mapping NAT ssh 22 to 22 don't worked for me. I used a different port for the host (2222).

File: docs/developers-docs/development-environment/frontend_environment.md
Comment (RaimiSol) [inline docs/developers-docs/development-environment/frontend_environment.md:110]: To get this work with default settings I used #{REST_URL} instead of http://#{$HOSTNAME}:8081

Comment (RaimiSol) [reply docs/developers-docs/development-environment/frontend_environment.md:110]: This applies also to the other lines involving http://#{$HOSTNAME}:8081

File: docs/developers-docs/development-environment/frontend_environment.md
Comment (RaimiSol) [inline docs/developers-docs/development-environment/frontend_environment.md:123]: I added $NOT_DOWNLOADABLE = []. Because if it is null it caused an error in ontologies_controller.

File: docs/developers-docs/development-environment/frontend_environment.md
Comment (RaimiSol) [inline docs/developers-docs/development-environment/frontend_environment.md:151]: I would also mention the one-time command:
docker compose exec web sh -c "bin/setup"

**Review by RaimiSol — CHANGES_REQUESTED** [review]

File: docs/developers-docs/development-environment/backend_environment.md
Comment (syphax-bouazzouni) [reply docs/developers-docs/development-environment/backend_environment.md:47]: I think it didn't work for you because had already an ssh server running on your machine. 
I will add a note about this. To tell to use port 2222 if the 22 didn't work.

**Comment by alexskr** [comment]
relying on the ontoportal appliance VM for running API for development purposes can be cumbersome.   So the question is should we make some effort to replace it with recently published docker-based API workflows or add that as a second option?

File: docs/developers-docs/development-environment/frontend_environment.md
Comment (syphax-bouazzouni) [reply docs/developers-docs/development-environment/frontend_environment.md:110]: Did you put your appliance IP in $HOSTAME instead of localhost

**Comment by syphax-bouazzouni** [comment]
Hi @alexskr.

Using the new docker-based environment would be, better.

I propose the following roadmap: 

1. First merge this first version (to not complexify this PR)
2. Secondly you can create another PR, to add as a recommended option to use the docker-based option (as a transition for teams/people that are already using the appliance-based env)
3. When we do the transition in our teams (and test it for our local code), I will do a last PR to remove the second option.

**Review by Bouchemel-Nasreddine — APPROVED** [review]

**Review by alexskr — APPROVED** [review]

**Pull request merged**
